### ISSUE TITLE
[CP] bump linter to fix crashes / records false positives

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#52185]
### Commit(s) to merge

92b963ec83b2b9a3ca40d08a67382d5a1644a37d

### Target

beta

### Prepared changelist for beta/stable

https://dart-review.googlesource.com/c/sdk/+/298741

### Issue Description

Fixes issues w/ patterns/records:

* crashes in `use_build_context_synchronously`
* false positives in `non_constant_identifier_names`
* false positives in `no_leading_underscores_for_local_identifiers`
* false positives in `constant_identifier_names`
* false positives in `prefer_final_locals`
 

### What is the fix

See linked fixes.

This bumps to the version of the linter currently in Google3.

We elected to do this rather than a series of targeted cherry picks because:

1. we moved to Dart 3.0 since the linter in stable making simple cherry picks hard and
2. an aggregate branch on the linter that collected the required changes would be complex

Since this linter is currently in Google3 and analytics have not shown any new crashes and we haven't received any new bug reports we feel this is safer than the alternative.

### Why cherry-pick

To fix crashes and fix recommended lint rule over-reporting for 3.0 code using patterns.

### Risk

low

### Issue link(s)

https://github.com/dart-lang/linter/pull/4279, https://github.com/dart-lang/linter/issues/4293, https://github.com/dart-lang/linter/issues/4295, https://github.com/dart-lang/linter/issues/4297, https://github.com/dart-lang/linter/issues/4299

### Extra Info

Full list of changes since version in `stable`:

e36813b3 fix `non_constant_identifier_names` pattern field over-reporting (https://github.com/dart-lang/linter/issues/4300)
70509465 fix `constant_identifier_names` pattern field over-reporting (https://github.com/dart-lang/linter/issues/4298)
ae3501be fix `no_leading_underscores_for_local_identifiers` pattern field over-reporting (v4296)
17dcb460 prefer_final_in_for_each_test for final records (https://github.com/dart-lang/linter/issues/4294)
f38c72ed tests for `prefer_final_in_for_each` pattern support (https://github.com/dart-lang/linter/issues/4291)
a7facd7d tidy up formal parameter inspection (https://github.com/dart-lang/linter/issues/4289)
0c18a657 `unnecessary_final` support for `ForEachPartsWithPattern`s (https://github.com/dart-lang/linter/issues/4288)
2212d95e fix `prefer_final_locals` false positive in foreach loops (https://github.com/dart-lang/linter/issues/4287)
6646b99d fix Dart3 unsafe mixing in (https://github.com/dart-lang/linter/issues/4285)
e48f2863 Move use_key_in_widget_constructors tests to be reflective (https://github.com/dart-lang/linter/issues/4283)
82e7147c Fix unrelated_type_equality_checks when comparing enums (https://github.com/dart-lang/linter/issues/4282)
d8ccf815 fix pr labeler yaqml indent (https://github.com/dart-lang/linter/issues/4051)
ca2a9fe0 Lint about comparing booleans to boolean literals (https://github.com/dart-lang/linter/issues/3973)
4c2af043 Fix false negative in `prefer_contains` (https://github.com/dart-lang/linter/issues/2830)
efa4c979 Update releasing.md with more pre-release steps (https://github.com/dart-lang/linter/issues/3901)
a8c903cb bump cli_util dep (https://github.com/dart-lang/linter/issues/4206)
500c2cec New rule: Use matching super parameter names (https://github.com/dart-lang/linter/issues/4263)
e22fdaee [use_build_context_synchronously] Fix for unprotected 'last' invocation. (https://github.com/dart-lang/linter/issues/4279)
f9edf665 bump gh pages build to dev SDK (https://github.com/dart-lang/linter/issues/4275)
e944fc70 hasLeadingUnderscore extension cleanup (https://github.com/dart-lang/linter/issues/4274)
676c7117 inverted token type extension (https://github.com/dart-lang/linter/issues/4272)
686e85da => Dart 3.0! (https://github.com/dart-lang/linter/issues/4267)
74fa5ff0 Add comments to use_build_context_synchronously tests with fuller explanations (https://github.com/dart-lang/linter/issues/4271)
a2155130 `prefer_const_constructors_in_immutables`: simplify visitConstructorDeclaration override (https://github.com/dart-lang/linter/issues/4254)
eb0ca501 Account for records in type relatedness checks (https://github.com/dart-lang/linter/issues/4266)
0b4787f7 stop pana baseline checking (https://github.com/dart-lang/linter/issues/4269)
d9c05ab5 Move more use_build_context_synchronously tests (https://github.com/dart-lang/linter/issues/4264)
fffb74c3 Move several if-concerned use_build_context_synchronously tests (https://github.com/dart-lang/linter/issues/4255)

Note that the unrelated changes were deemed low-risk.

**Comment by mit-mit** [comment]
SGTM

**Comment by itsjustkevin** [comment]
@pq in the case that this ends up causing issues, would there be any concerns with simply rolling back the revision?

**Comment by pq** [comment]
> in the case that this ends up causing issues, would there be any concerns with simply rolling back the revision?

Nope. I think that would be the right thing to do.  If there are issues we can just go back to what's in beta currently and tough it out until we can put together something for the next stable release.

**Comment by itsjustkevin** [comment]
@jacob314 and @vsmenon could you take a look at this cherry-pick request?

**Comment by srawlins** [comment]
I think another commit in the list is also a good reason to cherry-pick:

> eb0ca501 Account for records in type relatedness checks (https://github.com/dart-lang/linter/pull/4266)

**Comment by vsmenon** [comment]
@athomas @leafpetersen - thoughts?  We're getting close enough to the release - the bar is high.  Are we better off holding this for the next dot release?

**Comment by leafpetersen** [comment]
@pq can you summarize how common these issue are?  Is this fixing something that many people are likely to run into, or is this more of a "now and then" sort of thing?  We have a patch release going out ~1 week after stable, is it reasonable to push this out or is this something that's going to significantly impact the experience of a significant number of users when they try things out?

**Comment by pq** [comment]
@scheglov: can you speak to the frequency of the crash you fixed? (It's in a flutter recommended lint so could potentially affect a lot of 3P folks.)

A few of the false positives are in recommended lints and will trigger with records in particular though those are easier to (temporarily) ignore.

**Comment by scheglov** [comment]
High enough that I noticed it looking through crash reports.
I don't know enough to make good query, but probably using the existing crash UI [this](https://crash.corp.google.com/browse?q=product_name%3D%22Dart_analysis_server%22+AND+stable_signature%3D%27StateError+at+ListBase.last+at+dart%3Acollection-e699befa%27&stbtiq=&reportid=&index=0).
Currently it shows 751 crashes from 85 users that used beta.

**Comment by srawlins** [comment]
One more note: we definitely saw the crashes drop after the fix made it to the Dart SDK. Crashes in general are reduced as well (we did not create any worse problem with the fix :D )

**Comment by itsjustkevin** [comment]
@pq and @srawlins  assuming the failures here are expected we can merge this one.

**Comment by pq** [comment]
Yeah. Those failures are unrelated.
